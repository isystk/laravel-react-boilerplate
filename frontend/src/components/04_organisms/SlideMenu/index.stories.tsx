import { Meta, Story } from '@storybook/react'
import React from 'react'
import SlideMenu, { SlideMenuProps } from './index'

export default {
  title: '04_organisms/SlideMenu',
  component: SlideMenu,
} as Meta

const Template: Story = () => {
  const props: SlideMenuProps = {
    isMenuOpen: true,
    setMenuOpen: () => ({}),
    menuItems: [
      {
        label: 'Menu1',
        href: 'https://menu1',
      },
      {
        label: 'Menu2',
        href: 'https://menu2',
      },
      {
        label: 'Menu3',
        href: 'https://menu3',
        target: '_blank',
      },
    ],
  }
  return <SlideMenu {...props} />
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
