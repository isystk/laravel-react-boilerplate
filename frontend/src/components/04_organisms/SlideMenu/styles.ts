import { css, keyframes } from '@emotion/css'

export const slideMenu = css``

export const menuClose = css`
  width: 0;
  transition: 0.5s;
`
export const menuOpen = css`
  width: 70%;
`
