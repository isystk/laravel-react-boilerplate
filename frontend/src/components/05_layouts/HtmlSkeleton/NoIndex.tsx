import { connect } from '@/components/hoc'
import React from 'react'
import { ContainerProps } from 'types'

/** NoIndex Props */
export type NoIndexProps = Record<never, never>
export type PresenterProps = NoIndexProps

/** Presenter Component */
const Presenter: React.FC<PresenterProps> = () => <></>

/** Container Component */
const Container: React.FC<ContainerProps<NoIndexProps, PresenterProps>> = ({
  presenter,
  ...props
}) => {
  return presenter({ ...props })
}

/** NoIndex */
export default connect<NoIndexProps, PresenterProps>(
  'NoIndex',
  Presenter,
  Container
)
