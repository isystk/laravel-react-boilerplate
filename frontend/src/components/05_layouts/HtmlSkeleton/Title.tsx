import { connect } from '@/components/hoc'
import React, { Fragment } from 'react'
import { ContainerProps } from 'types'

/** Title Props */
export type TitleProps = { children }
export type PresenterProps = TitleProps

/** Presenter Component */
const Presenter: React.FC<PresenterProps> = ({ children }) => (
  <Fragment>{children}</Fragment>
)

/** Container Component */
const Container: React.FC<ContainerProps<TitleProps, PresenterProps>> = ({
  presenter,
  ...props
}) => {
  return presenter({ ...props })
}

/** Title */
export default connect<TitleProps, PresenterProps>(
  'Title',
  Presenter,
  Container
)
