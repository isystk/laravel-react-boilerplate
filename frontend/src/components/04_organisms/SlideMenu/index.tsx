import React, { FC } from 'react'
import { ContainerProps } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import { useI18n } from '@/components/i18n'
import { MenuItem } from '@/constants/menu'

/** SideMenuProps Props */
export type SideMenuProps = {
  setMenuOpen: () => void
  isMenuOpen: boolean
  menuItems: MenuItem[]
  position: 'right' | 'left'
}
/** Presenter Props */
export type PresenterProps = SideMenuProps & {
  setMenuOpen
  isMenuOpen
}

/** Presenter Component */
const SideMenuPresenter: FC<PresenterProps> = ({
  t,
  setMenuOpen,
  isMenuOpen,
  menuItems = [],
  position = 'right',
  ...props
}) => (
  <>
    <div
      className={`${styles.sideMenu} fixed top-0 ${
        position === 'left' ? 'left-0' : 'right-0'
      } z-30 bg-base h-screen w-64 ${
        isMenuOpen ? styles.menuOpen : styles.menuClose
      }`}
    >
      <div className="mt-24"></div>
      <ul className="list-none">
        {menuItems.map(({ label, href, target }, idx) => (
          <li className="py-4" key={idx}>
            <a
              href={href}
              target={target}
              rel={target ? 'noreferrer' : ''}
              className="block text-gray-700 font-bold whitespace-nowrap px-4"
            >
              {t(label)}
            </a>
          </li>
        ))}
      </ul>
    </div>
    <div
      className={`${
        isMenuOpen ? 'block' : 'hidden'
      } z-20 bg-black opacity-50 fixed top-0 w-full h-screen`}
      onClick={() => setMenuOpen(!isMenuOpen)}
    ></div>
  </>
)

/** Container Component */
const SideMenuContainer: React.FC<
  ContainerProps<SideMenuProps, PresenterProps>
> = ({ presenter, setMenuOpen, isMenuOpen, ...props }) => {
  const { t } = useI18n('Common')

  return presenter({
    t,
    setMenuOpen,
    isMenuOpen,
    ...props,
  })
}

export default connect<SideMenuProps, PresenterProps>(
  'SideMenu',
  SideMenuPresenter,
  SideMenuContainer
)
