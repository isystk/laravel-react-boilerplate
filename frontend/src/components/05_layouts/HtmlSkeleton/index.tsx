import React, { Children, FC } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import Title, { TitleProps } from './Title'
import NoIndex, { NoIndexProps } from './NoIndex'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { APP_NAME, APP_DESCRIPTION } from '@/constants'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: string
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  title,
  noIndex,
  description,
  children,
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* Index or Noindex */}
      <meta name="robots" content={noIndex ? 'noindex' : 'index'} />
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* PCやモバイル（スマホ、タブレット）などのデバイスごとのコンテンツの表示領域 */}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
    </Head>
    {children}
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  let noIndex = false

  children = Children.map(children, (child) => {
    if (isReactElement(child) && child.type === Title) {
      // Titleタグが含まれる場合
      return (title = `${child.props.children} | ${APP_NAME}`) && undefined
    }
    if (isReactElement(child) && child.type === NoIndex) {
      // NoIndexタグが含まれる場合
      return (noIndex = true) && undefined
    }
    return child
  })
  if (!title) title = APP_NAME

  const description = APP_DESCRIPTION
  return presenter({
    title,
    noIndex,
    description,
    children,
    ...props,
  })
}

export const Context = React.createContext()

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
export type { NoIndexProps }
export { NoIndex }
