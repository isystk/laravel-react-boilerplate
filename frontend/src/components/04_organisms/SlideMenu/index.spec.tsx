import React from 'react'
import renderer from 'react-test-renderer'
import SlideMenu, { SlideMenuProps } from './index'

import { Context } from '@/components/05_layouts/HtmlSkeleton'

describe('SlideMenu', () => {
  it('Match Snapshot', () => {
    const props: SlideMenuProps = {
      isMenuOpen: true,
      setMenuOpen: () => ({}),
      menuItems: [
        {
          label: 'Menu1',
          href: 'https://menu1',
        },
        {
          label: 'Menu2',
          href: 'https://menu2',
        },
        {
          label: 'Menu3',
          href: 'https://menu3',
          target: '_blank',
        },
      ],
    }

    const component = renderer.create(<SlideMenu {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
