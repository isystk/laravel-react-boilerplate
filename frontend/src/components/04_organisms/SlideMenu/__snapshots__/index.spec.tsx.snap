// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlideMenu Match Snapshot 1`] = `
[
  <div
    className="undefined fixed top-0 right-0 z-30 bg-base h-screen w-64 css-1mfqb2z"
  >
    <div
      className="mt-24"
    />
    <ul
      className="list-none"
    >
      <li
        className="py-4"
      >
        <a
          className="block text-gray-700 font-bold whitespace-nowrap px-4"
          href="https://menu1"
          rel=""
        >
          Menu1
        </a>
      </li>
      <li
        className="py-4"
      >
        <a
          className="block text-gray-700 font-bold whitespace-nowrap px-4"
          href="https://menu2"
          rel=""
        >
          Menu2
        </a>
      </li>
      <li
        className="py-4"
      >
        <a
          className="block text-gray-700 font-bold whitespace-nowrap px-4"
          href="https://menu3"
          rel="noreferrer"
          target="_blank"
        >
          Menu3
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="block z-20 bg-black opacity-50 fixed top-0 w-full h-screen"
    onClick={[Function]}
  />,
]
`;
